Makes the alert visible even if scrolled down

disissid opened the pull request:
Solves the issue #515 
- Makes the alert visible even if scrolled down by fixing it at top.
- Adds a cross button to close the warning if hindering the view

Screenshots-
Wider Screen
![screenshot from 2016-03-15 23 14 18](https://cloud.githubusercontent.com/assets/6582951/13787933/fdf28d4c-eb03-11e5-863c-6d1a044cba7c.png)

Smaller screen
![screenshot from 2016-03-15 23 14 44](https://cloud.githubusercontent.com/assets/6582951/13787946/0687422c-eb04-11e5-90a4-c305bf00736d.png)

smarx commented:
*Automated message from Dropbox CLA bot*

@disissid, thanks for the pull request! It looks like you haven't yet signed the Dropbox CLA. Please [sign it here](https://opensource.dropbox.com/cla/) and update the thread so we can consider merging your code.

disissid commented:
@timabbott Please review. And I'm not sure about this Travis error, I guess its a network error, right?

varshit97 commented:
@disissid Run "tools/lint-all" inside the vagrant and see if any errors come up. This might help with the TravisCI error. :)

timabbott commented on static/js/subs.js:
```diff
@@ -922,10 +930,12 @@ $(function () {
 
-                ui.report_success("The stream has been renamed!", $("#subscriptions-status"));
+                ui.report_success("The stream has been renamed!", $("#subscriptions-status "),
+                                'subscriptions-status');
```
This whitespace for this looks wrong...

timabbott commented on static/js/subs.js:
```diff
@@ -938,7 +948,7 @@ $(function () {
         var description = $sub_row.find('input[name="description"]').val();
 
-        $('#subscription_status').hide();
+        $('#subscriptions-status').hide();
```
Hmm, I guess the previous code here for hiding errors was just buggy?

disissid replied:
Yes, it was.

timabbott commented on static/styles/zulip.css:
```diff
@@ -3895,6 +3915,11 @@ form.admin-realm .control-label {
 
+    #subscriptions-status {
+    left:250px;
+    width:50%;
```
The whitespace here looks wrong.

timabbott commented on static/styles/zulip.css:
```diff
@@ -3981,6 +4006,11 @@ form.admin-realm .control-label {
 
+    #subscriptions-status {
+    left:0px;
+    width:70%;
```
here too

timabbott commented on static/js/ui.js:
```diff
@@ -122,28 +122,38 @@ function message_hover(message_row) {
+        status_box.removeClass(status_classes).addClass(cls)
               .text(response).stop(true).fadeTo(0, 1);
+    }
+
```
Can you explain why you needed to add an additional argument here?  I'm not sure I see why this couldn't have used the existing framework...

disissid replied:
I added elements to the div(#subscriptions-status)
`  <div class="alert" id="subscriptions-status">
      <span id="response"></span>
      <span id="close-subscriptions-status"><i class="icon-vector-remove"></i></span>            </div>`

And .text() replaces all the content of an element even including child elements, so without messing with the earlier code I checked whether the call is coming from subscriptons-status or not, if yes then I changed the content(..text()) of only '#response' span.

I could've done it with the existing framework but then I would've to show and hide 'remove' button every time explicitly. Do you dee a better approach here?

timabbott replied:
hmm, OK, I guess this is a reasonable solution.  I'm a bit worried that the optional argument will lead to problems with people confusing the arguments in the future, but I guess we can always try to improve these APIs later.

disissid replied:
Yeah that can confuse people, should I add some comments there for now?

timabbott replied:
Yeah, I think we're probably at the point where the report_message/report_error/report_success tree of functions could benefit from a block comment explaining what the arguments mean at the top.

timabbott replied:
@disissid can you either add comments here to this PR or open an issue for adding that later?

disissid replied:
@timabbott I'll add the comments to this PR only.

timabbott commented:
@disissid sure looks like a network error; Travis CI was having a lot of those for a few days.  

Thanks for working on this!  I posted a bunch of comments; if you can fix those before the weekend I can do a complete review and maybe merge this then.

disissid commented:
@timabbott Fixed them, review if anything else need to be changed.

timabbott commented on static/styles/zulip.css:
```diff
@@ -3896,6 +3916,11 @@ form.admin-realm .control-label {
 
+    #subscriptions-status {
+        left:250px;
+        width:50%;
```
these still don't have correct whitespace; there should be a spaec between the : and the number. (issue is still present below too)

disissid replied:
Oh yeah, Sorry didn't notice them.Fixed now.

timabbott commented:
Looks like there's an issue with the error message bar with a very wide window (this was taken on a 24" monitor):

![image](https://cloud.githubusercontent.com/assets/2746074/13900726/64921d06-edca-11e5-9044-06f11c490bff.png)

I think how this should work from a design perspective is that the error bar should be inside the center pane, and ideally would be setup in a way to not overlap with the "Streams" heading for the top of the page...

disissid commented:
For overlapping the "Streams" heading, div is fixed so it is positioned relative to viewport, To fix it I 
increased the bottom margin of "Streams" heading which doesn't look bad without warning 

![screenshot from 2016-03-20 03 54 26](https://cloud.githubusercontent.com/assets/6582951/13901630/3fb6ae26-ee50-11e5-9e4f-46cbee3b1b4e.png)

With warning 
![screenshot from 2016-03-20 03 54 37](https://cloud.githubusercontent.com/assets/6582951/13901635/4d96cb66-ee50-11e5-9e55-89f3f72ba6d3.png)

This can overlap other stream input div if warning would extend to next line but that would be very rare case.
I could've checked if user has scrolled down or not but that would be unnecessarily keeping a check for event that would happen rarely.(I suppose with this approach I can debounce the function)

timabbott commented:
OK, that's reasonable for that piece.

disissid commented:
@timabbott Done.

timabbott commented:
Hmm, I'd prefer the issue for larger streams be fixed by just making the error bar a part of the middle panel, rather than adding this sort of CSS based on the current size of the elements in the middle panel -- can you do something of that form instead?

disissid commented:
Ah, again the error bar's position is fixed now, which is positioned relative to the viewport. BTW what is wrong with this approach?

timabbott commented:
The problem with this approach is that it's complex (in terms of how much code is involved) and somewhat brittle -- if we were to change the design to make the center panel 20% larger, this code would break.  Whereas if the error bar element just had the width of the center panel by construction, it'd be less code and not subject to that issue.

disissid commented:
Okay, that makes sense, I had to change explicitly for different screen sizes. Now I've set its width with respect to the middle column. Please review @timabbott.

timabbott commented on static/js/subs.js:
```diff
@@ -1117,8 +1117,23 @@ $(function () {
+        set_subscription_status_width();
+    });
+
+    set_subscription_status_width();
```
Would it work to do this in pure CSS by making the templates/zerver/subscriptions.html  `<div class="subscriptions">` container div have the `.column-middle` class (or something similar)?  This still seems like it's more complicated than the "an element inside the larger middle column element" should need to be...

disissid replied:
I'm not sure how that would be possible having a fixed element and setting its dimensions with respect to a static element without the use of javascript , one possible way would have been to inherit the width of parent element(in css) but again it is dynamic too(middle column, for this to work width of parent should be defined explicitly in css).

Ah, Let me see if I can do something else.

disissid replied:
Found it, `app-main` has max-width of 1400px, changed it with respect to it.Made it so simple :)

disissid commented:
@timabbott What do you think now?

timabbott commented:
Great, that's roughly what I had in mind!  Two things:
* You should squash the commits into a single commit to match http://zulip.readthedocs.org/en/latest/code-style.html#version-control
* I tried testing this by loading my development environment, using ctrl-C to kill the dev server, and then clicking some subscribe/unsubscibe buttons.  If you're scrolled down in the page when you click a subscribe/unsubscribe button, it scrolls you to the top (I assume with the intent of showing you the error bar), which is no longer necessary and a bit confusing since it obscures what it was that you actually clicked.  Can you look into fixing this as part of this change?

disissid commented:
@timabbott Squashed the commits and fixed the issue. Actually it was scrolling up to focus on stream input but I removed it anyway as you said it was confusing since it obscures what it was that you actually clicked.

timabbott commented:
OK, I tested this around for a bit, and I think it works correctly; it's be good to address the `report_*` comment issue.  Also, @disissid can you fix the commit message?  It shouldn't just have a list of the commit messages of all your various test commits, but instead a description of what a developer seeing this change might want to know to understand why it is correct.

Also, the commit message should have a sentence of the form `Fixes: #515.` so it automatically closes the original issue when it gets merged. 

@wdaher do you want to take a quick look at the CSS here?  

Here's a screenshot:
![image](https://cloud.githubusercontent.com/assets/2746074/14071320/365822f6-f468-11e5-92c7-ce560c30b215.png)

I feel like we could make this look better, but I'm not sure what the best approach is.  Maybe move the error message to sit below the search box rather than hovering over the scrollable pane instead?

wdaher commented:
My view is that the screenshot you show, Tim, is better than nothing but
pretty ugly.

>  Maybe move the error message to sit below the search box rather than
hovering over the scrollable pane instead?

This seems promising to me. What do we do today when you lose your network
connection, and can we borrow from that logic?

On Sun, Mar 27, 2016 at 10:08 PM, Tim Abbott <notifications@github.com>
wrote:

> OK, I tested this around for a bit, and I think it works correctly; it's
> be good to address the report_* comment issue. Also, @disissid
> <https://github.com/disissid> can you fix the commit message? It
> shouldn't just have a list of the commit messages of all your various test
> commits, but instead a description of what a developer seeing this change
> might want to know to understand why it is correct.
>
> Also, the commit message should have a sentence of the form Fixes: #515.
> so it automatically closes the original issue when it gets merged.
>
> @wdaher <https://github.com/wdaher> do you want to take a quick look at
> the CSS here?
>
> Here's a screenshot:
> [image: image]
> <https://cloud.githubusercontent.com/assets/2746074/14071320/365822f6-f468-11e5-92c7-ce560c30b215.png>
>
> I feel like we could make this look better, but I'm not sure what the best
> approach is. Maybe move the error message to sit below the search box
> rather than hovering over the scrollable pane instead?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zulip/zulip/pull/534#issuecomment-202236303>
>

timabbott commented:
When you lose your network connection, we display this in the buddy list sidebar:

![image](https://cloud.githubusercontent.com/assets/2746074/14071634/dd4a3dbc-f46b-11e5-8134-367ea16e8a55.png)

(there's 2 possible types of errors, and it appears you get both here in the unusual case that it loses the connection while the page is loading; which probably could use improvement in its own right).  But that's for displaying an error with the overall state as opposed to one caused by a particular UI widget; most of our UI errors display as a text bar at the top of the relevant widget.

wdaher commented:
Yeah, that's not good enough in this case, agreed.

On Sun, Mar 27, 2016 at 10:35 PM, Tim Abbott <notifications@github.com>
wrote:

> When you lose your network connection, we display this in the buddy list
> sidebar:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/2746074/14071634/dd4a3dbc-f46b-11e5-8134-367ea16e8a55.png>
>
> (there's 2 possible types of errors, and it appears you get both here in
> the unusual case that it loses the connection while the page is loading;
> which probably could use improvement in its own right). But that's for
> displaying an error with the overall state as opposed to one caused by a
> particular UI widget; most of our UI errors display as a text bar at the
> top of the relevant widget.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zulip/zulip/pull/534#issuecomment-202241816>
>

timabbott commented:
OK, @disissid, are you up for making an attempt at having the error message sit below the search box instead?

disissid commented:
Yeah, @timabbott I'll do it and update the PR by tomorrow.

disissid commented:
@wdaher @timabbott  What do you think now?

![screenshot from 2016-04-03 21 27 42](https://cloud.githubusercontent.com/assets/6582951/14233717/f659208e-f9ed-11e5-9019-9d712709d0f8.png)


![screenshot from 2016-04-03 21 28 00](https://cloud.githubusercontent.com/assets/6582951/14233718/f68921c6-f9ed-11e5-8cde-f5bb68ba154b.png)

wdaher commented:
Looks good to me.

Two comments:

1) Can you vertically center the 'x' where it is? Right now it looks like
it's floating a bit towards the top, which looks kind of funny to me.

2) In screenshot 2, if you click the 'x', does it actually click the x and
stop or does it trigger the click of the 'v'?

On Sun, Apr 3, 2016 at 10:17 AM, Kumar Siddharth <notifications@github.com>
wrote:

> @wdaher <https://github.com/wdaher> @timabbott
> <https://github.com/timabbott> What do you think now?
>
> [image: screenshot from 2016-04-03 21 27 42]
> <https://cloud.githubusercontent.com/assets/6582951/14233717/f659208e-f9ed-11e5-9019-9d712709d0f8.png>
>
> [image: screenshot from 2016-04-03 21 28 00]
> <https://cloud.githubusercontent.com/assets/6582951/14233718/f68921c6-f9ed-11e5-8cde-f5bb68ba154b.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zulip/zulip/pull/534#issuecomment-205013575>
>

disissid commented:
Hi @wdaher 

1) That was intentional, I agree it doesn't look good when error is of one line but for narrow screens I don't think we would want to center the 'x' vertically.

![screenshot from 2016-04-04 13 10 38](https://cloud.githubusercontent.com/assets/6582951/14241079/3bc11c8e-fa67-11e5-9726-585f4024386f.png)

Ah, I can center it vertically for only wider screens.

2) Clicking the 'x' won't trigger the 'v'.

wdaher commented:
Great point, in the multi line case you definitely don't want to center it.
I am fine with either selectively centering when it's one line using CSS,
or just leaving it as written, whatever feels most consistent with what we
do elsewhere in the app.

Thanks!
On Mon, Apr 4, 2016 at 12:47 AM Kumar Siddharth <notifications@github.com>
wrote:

> Hi @wdaher <https://github.com/wdaher>
>
> 1) That was intentional, I agree it doesn't look good when error is of one
> line but for narrow screens I don't think we would want to center the 'x'
> vertically.
>
> [image: screenshot from 2016-04-04 13 10 38]
> <https://cloud.githubusercontent.com/assets/6582951/14241079/3bc11c8e-fa67-11e5-9726-585f4024386f.png>
>
> Ah, I can center it vertically for only wider screens.
>
> 2) Clicking the 'x' won't trigger the 'v'.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/zulip/zulip/pull/534#issuecomment-205178169>
>

timabbott commented on static/js/subs.js:
```diff
@@ -558,13 +559,13 @@ function ajaxSubscribe(stream) {
-            ui.report_error("Error adding subscription", xhr, $("#subscriptions-status"));
-            $("#create_stream_name").focus();
+            ui.report_error("Error adding subscription", xhr,
+                            $("#subscriptions-status"), 'subscriptions-status');
```
these tabs should be spaces (same in the next block)

disissid replied:
Ah, I tried to change them, but git didn't recognize any change in the file.

timabbott commented:
To address @wdaher's point, @disissid what do you think about tweaking the CSS for the close button to be placed at an offset from the top that would be centered if it takes up 1 line but not if it takes up 2?

timabbott commented on templates/zerver/subscriptions.html:
```diff
@@ -3,7 +3,10 @@
-      <div class="alert" id="subscriptions-status"></div>
+	    <div class="alert" id="subscriptions-status">
+	      <span id="response"></span>
+	      <span id="close-subscriptions-status"><i class="icon-vector-remove"></i></span>
```
This block seems indented more than it should be -- I think the `subscription-status` div should be aligned with the <h1>

timabbott commented:
Otherwise this looks great!  @disissid ping me once you've addressed those and I'll merge.  

Thanks for working through making this look good @disissid !

disissid commented:
@wdaher @timabbott I won't suggest centering the close button as its height depends on the content of `subscritptions-status` div, so it would require unnecessary logic to see when the height of the div is changed and anyway current UI doesn't look that bad to me.

timabbott commented:
OK, I did a small bit of cleanup (e.g. cleaning up the commit message to match our commit messge style and fixing a few whitespace issues) and merged this as 4eced6922832d0494b30709d2566790b6db80a23.  Thanks for building this and tweaking the details of get this right @disissid!
